segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -502,32 +502,27 @@ public void commitTransaction(CommitTransaction commitTx) {
-        // but invoking them in parallel does provide benefits in terms of reduced latency.
-        val seal = segmentStore
-                .sealStreamSegment(transactionName, TIMEOUT)
-                .exceptionally(this::ignoreSegmentSealed)
-                .thenCompose(v -> recordStatForTransaction(transactionName, commitTx.getSegment())
+        // TODO: change SegmentStore to auto-seal before merging (https://github.com/pravega/pravega/issues/2558)
+        String transactionName = StreamSegmentNameUtils.getTransactionNameFromId(commitTx.getSegment(), commitTx.getTxid());
+        segmentStore.sealStreamSegment(transactionName, TIMEOUT)
+                    .exceptionally(this::ignoreSegmentSealed)
+                    .thenCompose(v -> recordStatForTransaction(transactionName, commitTx.getSegment())
                         .exceptionally(e -> {
                             // gobble up any errors from stat recording so we do not affect rest of the flow.
-                            log.error("exception while computing stats while merging txn {}", e);
+                            log.error("exception while computing stats while merging txn {}", commitTx, e);
                             return null;
-                        }));
-        val merge = segmentStore
-                .mergeTransaction(transactionName, TIMEOUT)
-                .thenAccept(v -> connection.send(new TransactionCommitted(requestId, commitTx.getSegment(), commitTx.getTxid())));
-
-        CompletableFuture.allOf(seal, merge)
-                .exceptionally(e -> {
-                    if (Exceptions.unwrap(e) instanceof StreamSegmentMergedException) {
-                        log.info("Stream segment is already merged '{}'.", transactionName);
-                        connection.send(new TransactionCommitted(requestId, commitTx.getSegment(), commitTx.getTxid()));
-                        return null;
-                    } else {
-                        return handleException(requestId, transactionName, "Commit transaction", e);
-                    }
-                });
+                        }))
+                    .thenCompose(v -> segmentStore.mergeStreamSegment(commitTx.getSegment(), transactionName, TIMEOUT))
+                    .thenAccept(v -> connection.send(new TransactionCommitted(requestId, commitTx.getSegment(), commitTx.getTxid())))
+                    .exceptionally(e -> {
+                        if (Exceptions.unwrap(e) instanceof StreamSegmentMergedException) {
+                            log.info("Stream segment is already merged '{}'.", transactionName);
+                            connection.send(new TransactionCommitted(requestId, commitTx.getSegment(), commitTx.getTxid()));
+                            return null;
+                        } else {
+                            return handleException(requestId, transactionName, "Commit transaction", e);
```

Comment:
is there a way to distinguish between errors thrown for source vs target segment?

Reply:
most exceptions here contain the name of the Segment which is at fault. However, that is only passed up via the message, and AFAIK, there is no transmission of any such information or stack trace over the wire.